File: typescript/tests/schema.test.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import { setupTelemetry, _resetConfig } from '../src/config';
import {
  EventSchemaError,
  event,
  eventName,
  getStrictSchema,
  setStrictSchema,
  validateEventName,
  validateRequiredKeys,
  _resetStrictSchemaForTests,
} from '../src/schema';
import { TelemetryError } from '../src/exceptions';

describe('setStrictSchema / getStrictSchema', () => {
  afterEach(() => {
    _resetStrictSchemaForTests();
    _resetConfig();
  });

  it('getStrictSchema returns config value when no override is set', () => {
    _resetStrictSchemaForTests();
    setupTelemetry({ strictSchema: false });
    expect(getStrictSchema()).toBe(false);
    _resetConfig();
    setupTelemetry({ strictSchema: true });
    expect(getStrictSchema()).toBe(true);
  });

  it('setStrictSchema overrides config and getStrictSchema returns override', () => {
    setupTelemetry({ strictSchema: false });
    setStrictSchema(true);
    expect(getStrictSchema()).toBe(true);
  });

  it('setStrictSchema(false) override takes precedence over config true', () => {
    setupTelemetry({ strictSchema: true });
    setStrictSchema(false);
    expect(getStrictSchema()).toBe(false);
  });

  it('_resetStrictSchemaForTests clears the override', () => {
    setStrictSchema(true);
    _resetStrictSchemaForTests();
    // After reset, falls back to config; default config has strictSchema: false
    _resetConfig();
    expect(getStrictSchema()).toBe(false);
  });

  it('setStrictSchema(true) causes event() to reject non-conforming segments', () => {
    setStrictSchema(true);
    expect(() => event('Auth', 'login', 'success')).toThrow(EventSchemaError);
  });

  it('setStrictSchema(false) causes event() to accept non-conforming segments', () => {
    setStrictSchema(false);
    const rec = event('Auth', 'Login', 'Success');
    expect(rec.event).toBe('Auth.Login.Success');
  });
});

describe('EventSchemaError', () => {
  it('is a TelemetryError', () => {
    const e = new EventSchemaError('bad');
    expect(e).toBeInstanceOf(TelemetryError);
    expect(e.name).toBe('EventSchemaError');
  });
});

describe('eventName', () => {
  beforeEach(() => {
    setupTelemetry({ strictSchema: true });
  });

  afterEach(() => {
    _resetConfig();
  });

  it('joins valid segments with dots', () => {
    expect(eventName('app', 'user', 'login')).toBe('app.user.login');
    expect(eventName('service', 'api', 'request', 'ok')).toBe('service.api.request.ok');
    expect(eventName('a', 'b', 'c', 'd', 'e')).toBe('a.b.c.d.e');
  });

  it('throws with too few segments', () => {
    expect(() => eventName('a', 'b')).toThrow(EventSchemaError);
    expect(() => eventName('a')).toThrow(EventSchemaError);
    expect(() => eventName()).toThrow(EventSchemaError);
  });

  it('throws with too many segments', () => {
    expect(() => eventName('a', 'b', 'c', 'd', 'e', 'f')).toThrow(EventSchemaError);
  });

  it('throws with invalid segment characters', () => {
    expect(() => eventName('App', 'user', 'login')).toThrow(EventSchemaError); // uppercase
    expect(() => eventName('app', 'user-login', 'event')).toThrow(EventSchemaError); // hyphen
    expect(() => eventName('0app', 'user', 'login')).toThrow(EventSchemaError); // starts with digit
    expect(() => eventName('app user', 'x', 'y')).toThrow(EventSchemaError); // space
  });

  it('allows digits and underscores in non-first position', () => {
    expect(eventName('app2', 'user_v2', 'login_ok')).toBe('app2.user_v2.login_ok');
  });
});

describe('validateEventName', () => {
  it('passes valid strict names', () => {
    expect(() => validateEventName('app.user.login')).not.toThrow();
    expect(() => validateEventName('a.b.c.d.e')).not.toThrow();
  });

  it('throws for invalid strict names', () => {
    expect(() => validateEventName('app.user')).toThrow(EventSchemaError); // only 2 segments
    expect(() => validateEventName('App.user.login')).toThrow(EventSchemaError); // uppercase
    expect(() => validateEventName('a.b.c.d.e.f')).toThrow(EventSchemaError); // 6 segments
    expect(() => validateEventName('app..login')).toThrow(EventSchemaError); // empty segment
  });

  it('relaxed mode accepts any non-empty segments', () => {
    expect(() => validateEventName('Foo.Bar', false)).not.toThrow();
    expect(() => validateEventName('one', false)).not.toThrow();
    expect(() => validateEventName('a.b.c.d.e.f.g', false)).not.toThrow();
  });

  it('relaxed mode rejects empty segments', () => {
    expect(() => validateEventName('a..b', false)).toThrow(EventSchemaError);
    expect(() => validateEventName('', false)).toThrow(EventSchemaError);
  });
});

describe('validateRequiredKeys', () => {
  it('passes when all keys are present', () => {
    expect(() => validateRequiredKeys({ a: 1, b: 2, c: 3 }, ['a', 'b'])).not.toThrow();
  });

  it('throws listing missing keys (sorted)', () => {
    const err = (() => {
      try {
        validateRequiredKeys({ a: 1 }, ['z', 'b', 'a']);
      } catch (e) {
        return e;
      }
    })();
    expect(err).toBeInstanceOf(EventSchemaError);
    expect((err as EventSchemaError).message).toContain('b');
    expect((err as EventSchemaError).message).toContain('z');
  });

  it('passes with empty required keys', () => {
    expect(() => validateRequiredKeys({}, [])).not.toThrow();
  });
});

describe('eventName / validateEventName — error message content (mutation kills)', () => {
  beforeEach(() => {
    setupTelemetry({ strictSchema: true });
  });

  afterEach(() => {
    _resetConfig();
  });

  it('segment count error mentions segment count', () => {
    // Kills: StringLiteral mutation that empties the error message
    expect(() => eventName('a', 'b')).toThrow(/got 2/);
    expect(() => eventName('a', 'b', 'c', 'd', 'e', 'f')).toThrow(/got 6/);
  });

  it('invalid segment error mentions the segment index', () => {
    // Kills: StringLiteral mutation that empties the segment error
    expect(() => eventName('valid', 'INVALID', 'seg')).toThrow(/segment\[1\]/);
  });

  it('validateEventName relaxed mode error mentions segment', () => {
    expect(() => validateEventName('valid..empty', false)).toThrow(/segment/i);
  });
});

describe('validateRequiredKeys — error message content (mutation kills)', () => {
  it('lists missing keys in sorted order', () => {
    // Kills: StringLiteral + sort-removal mutations
    const obj = { a: 1 };
    expect(() => validateRequiredKeys(obj, ['z', 'a', 'm'])).toThrow(/m, z/); // sorted, only missing ones
  });
});

describe('validateEventName — strict mode error message content (kills StringLiteral at schema.ts:52)', () => {
  it('error message mentions segment count when too few segments', () => {
    // Kills: StringLiteral mutation that empties validateEventName strict-mode error message
    let msg = '';
    try {
      validateEventName('app.user');
    } catch (e) {
      msg = (e as Error).message;
    }
    expect(msg).toMatch(/expected 3-5 segments/);
    expect(msg).toMatch(/got 2/);
  });

  it('error message mentions segment count when too many segments', () => {
    let msg = '';
    try {
      validateEventName('a.b.c.d.e.f');
    } catch (e) {
      msg = (e as Error).message;
    }
    expect(msg).toMatch(/expected 3-5 segments/);
    expect(msg).toMatch(/got 6/);
  });
});

describe('eventName — join separator is dot (kills .join mutation at line 29)', () => {
  beforeEach(() => {
    setupTelemetry({ strictSchema: true });
  });

  afterEach(() => {
    _resetConfig();
  });

  it('segments are joined with dots, not other separators', () => {
    const name = eventName('app', 'user', 'login');
    expect(name).toBe('app.user.login');
    // Verify dot is the separator specifically
    expect(name).toContain('.');
    expect(name.split('.').length).toBe(3);
    expect(name.split('.')[0]).toBe('app');
    expect(name.split('.')[1]).toBe('user');
    expect(name.split('.')[2]).toBe('login');
  });

  it('result contains dots between every pair of segments', () => {
    const name = eventName('a', 'b', 'c', 'd', 'e');
    // Verify every adjacent pair is separated by exactly '.'
    expect(name).toBe('a.b.c.d.e');
    expect(name.indexOf('.')).toBe(1);
    expect(name.charAt(1)).toBe('.');
    expect(name.charAt(3)).toBe('.');
  });
});

describe('eventName — strict schema config integration', () => {
  afterEach(() => {
    _resetConfig();
  });

  it('relaxed mode allows 1 segment', () => {
    setupTelemetry({ strictSchema: false });
    expect(eventName('svc')).toBe('svc');
  });

  it('relaxed mode allows 6 segments', () => {
    setupTelemetry({ strictSchema: false });
    expect(eventName('a', 'b', 'c', 'd', 'e', 'f')).toBe('a.b.c.d.e.f');
  });

  it('strict mode rejects 1 segment', () => {
    setupTelemetry({ strictSchema: true });
    expect(() => eventName('svc')).toThrow(EventSchemaError);
  });

  it('0 segments throws even in relaxed mode', () => {
    setupTelemetry({ strictSchema: false });
    expect(() => eventName()).toThrow(EventSchemaError);
  });
});

describe('event()', () => {
  beforeEach(() => {
    setupTelemetry({ strictSchema: true });
  });

  afterEach(() => {
    _resetConfig();
  });

  it('returns structured EventRecord for 3 segments (DAS)', () => {
    const rec = event('auth', 'login', 'success');
    expect(rec).toEqual({
      event: 'auth.login.success',
      domain: 'auth',
      action: 'login',
      status: 'success',
    });
    expect(rec.resource).toBeUndefined();
  });

  it('returns structured EventRecord for 4 segments (DARS)', () => {
    const rec = event('db', 'query', 'orders', 'failure');
    expect(rec).toEqual({
      event: 'db.query.orders.failure',
      domain: 'db',
      action: 'query',
      resource: 'orders',
      status: 'failure',
    });
  });

  it('throws EventSchemaError with 2 segments', () => {
    expect(() => event('a', 'b')).toThrow(EventSchemaError);
    expect(() => event('a', 'b')).toThrow(/requires 3 or 4 segments/);
  });

  it('throws EventSchemaError with 5 segments', () => {
    expect(() => event('a', 'b', 'c', 'd', 'e')).toThrow(EventSchemaError);
    expect(() => event('a', 'b', 'c', 'd', 'e')).toThrow(/requires 3 or 4 segments/);
  });

  it('throws EventSchemaError with 0 segments', () => {
    expect(() => event()).toThrow(EventSchemaError);
  });

  it('throws EventSchemaError with 1 segment', () => {
    expect(() => event('a')).toThrow(EventSchemaError);
  });

  it('strict mode validates segment format', () => {
    expect(() => event('Auth', 'login', 'success')).toThrow(EventSchemaError);
    expect(() => event('auth', 'LOGIN', 'success')).toThrow(/does not match pattern/);
    expect(() => event('auth', 'login', '0bad')).toThrow(EventSchemaError);
  });

  it('relaxed mode skips segment format validation', () => {
    _resetConfig();
    setupTelemetry({ strictSchema: false });
    const rec = event('Auth', 'Login', 'Success');
    expect(rec.event).toBe('Auth.Login.Success');
    expect(rec.domain).toBe('Auth');
  });

  it('eventName() still works as deprecated alias', () => {
    const name = eventName('auth', 'login', 'success');
    expect(name).toBe('auth.login.success');
    expect(typeof name).toBe('string');
  });
});

File: go/schema_test.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

package telemetry

import (
	"errors"
	"strings"
	"testing"
)

func TestEventName_Valid(t *testing.T) {
	cases := []struct {
		name     string
		segments []string
		wantName string
	}{
		{
			name:     "exactly 3 segments",
			segments: []string{"provide", "telemetry", "setup"},
			wantName: "provide.telemetry.setup",
		},
		{
			name:     "exactly 5 segments",
			segments: []string{"provide", "telemetry", "setup", "complete", "ok"},
			wantName: "provide.telemetry.setup.complete.ok",
		},
		{
			name:     "4 segments",
			segments: []string{"provide", "telemetry", "setup", "done"},
			wantName: "provide.telemetry.setup.done",
		},
		{
			name:     "segment with digits and underscores",
			segments: []string{"a1", "b2_c3", "d4e5"},
			wantName: "a1.b2_c3.d4e5",
		},
		{
			name:     "segment starts with letter followed by underscore",
			segments: []string{"my_module", "sub_component", "event_fired"},
			wantName: "my_module.sub_component.event_fired",
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got, err := EventName(tc.segments...)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if got != tc.wantName {
				t.Errorf("want %q, got %q", tc.wantName, got)
			}
		})
	}
}

func TestEventName_Invalid(t *testing.T) {
	// Format validation (segment content) requires strict mode.
	_strictSchema = true
	t.Cleanup(func() { _strictSchema = false })

	cases := []struct {
		name        string
		segments    []string
		errContains string
	}{
		{
			name:        "too few segments (2)",
			segments:    []string{"provide", "telemetry"},
			errContains: "3",
		},
		{
			name:        "too many segments (6)",
			segments:    []string{"a", "b", "c", "d", "e", "f"},
			errContains: "6",
		},
		{
			name:        "uppercase in segment",
			segments:    []string{"Provide", "telemetry", "setup"},
			errContains: "Provide",
		},
		{
			name:        "segment starting with digit",
			segments:    []string{"1provide", "telemetry", "setup"},
			errContains: "1provide",
		},
		{
			name:        "empty segment",
			segments:    []string{"", "telemetry", "setup"},
			errContains: "",
		},
		{
			name:        "segment with hyphen",
			segments:    []string{"provide-io", "telemetry", "setup"},
			errContains: "provide-io",
		},
		{
			name:        "segment with space",
			segments:    []string{"provide io", "telemetry", "setup"},
			errContains: "provide io",
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got, err := EventName(tc.segments...)
			if err == nil {
				t.Fatalf("expected error, got name %q", got)
			}
			var schemaErr *EventSchemaError
			if !errors.As(err, &schemaErr) {
				t.Errorf("expected *EventSchemaError, got %T: %v", err, err)
			}
			if tc.errContains != "" && !strings.Contains(err.Error(), tc.errContains) {
				t.Errorf("error %q should contain %q", err.Error(), tc.errContains)
			}
			if got != "" {
				t.Errorf("on error expected empty name, got %q", got)
			}
		})
	}
}

func TestEventName_ZeroSegments(t *testing.T) {
	_, err := EventName()
	if err == nil {
		t.Fatal("expected error for zero segments")
	}
	var schemaErr *EventSchemaError
	if !errors.As(err, &schemaErr) {
		t.Errorf("expected *EventSchemaError, got %T", err)
	}
}

func TestValidateEventName_Valid(t *testing.T) {
	cases := []string{
		"provide.telemetry.setup",
		"provide.telemetry.setup.done",
		"provide.telemetry.setup.complete.ok",
	}
	for _, name := range cases {
		t.Run(name, func(t *testing.T) {
			if err := ValidateEventName(name); err != nil {
				t.Errorf("unexpected error for %q: %v", name, err)
			}
		})
	}
}

func TestValidateEventName_Invalid(t *testing.T) {
	// Format validation (segment content) requires strict mode.
	_strictSchema = true
	t.Cleanup(func() { _strictSchema = false })

	cases := []struct {
		name        string
		errContains string
	}{
		{"too.short", "2"},
		{"a.b.c.d.e.f", "6"},
		{"Provide.telemetry.setup", "Provide"},
		{"1start.telemetry.setup", "1start"},
		{"provide.telemetry.bad-seg", "bad-seg"},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			err := ValidateEventName(tc.name)
			if err == nil {
				t.Fatalf("expected error for %q", tc.name)
			}
			var schemaErr *EventSchemaError
			if !errors.As(err, &schemaErr) {
				t.Errorf("expected *EventSchemaError, got %T: %v", err, err)
			}
			if tc.errContains != "" && !strings.Contains(err.Error(), tc.errContains) {
				t.Errorf("error %q should contain %q", err.Error(), tc.errContains)
			}
		})
	}
}

func TestValidateEventName_RoundTrip(t *testing.T) {
	segments := []string{"provide", "telemetry", "event"}
	name, err := EventName(segments...)
	if err != nil {
		t.Fatalf("EventName: %v", err)
	}
	if err := ValidateEventName(name); err != nil {
		t.Errorf("ValidateEventName round-trip failed: %v", err)
	}
}

func TestEvent_DAS(t *testing.T) {
	evt, err := Event("auth", "login", "success")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if evt.Event != "auth.login.success" {
		t.Errorf("want Event=%q, got %q", "auth.login.success", evt.Event)
	}
	if evt.Domain != "auth" {
		t.Errorf("want Domain=%q, got %q", "auth", evt.Domain)
	}
	if evt.Action != "login" {
		t.Errorf("want Action=%q, got %q", "login", evt.Action)
	}
	if evt.Resource != "" {
		t.Errorf("want empty Resource, got %q", evt.Resource)
	}
	if evt.Status != "success" {
		t.Errorf("want Status=%q, got %q", "success", evt.Status)
	}
}

func TestEvent_DARS(t *testing.T) {
	evt, err := Event("payment", "subscription", "renewal", "success")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if evt.Event != "payment.subscription.renewal.success" {
		t.Errorf("want Event=%q, got %q", "payment.subscription.renewal.success", evt.Event)
	}
	if evt.Domain != "payment" {
		t.Errorf("want Domain=%q, got %q", "payment", evt.Domain)
	}
	if evt.Action != "subscription" {
		t.Errorf("want Action=%q, got %q", "subscription", evt.Action)
	}
	if evt.Resource != "renewal" {
		t.Errorf("want Resource=%q, got %q", "renewal", evt.Resource)
	}
	if evt.Status != "success" {
		t.Errorf("want Status=%q, got %q", "success", evt.Status)
	}
}

func TestEvent_TwoArgs_Error(t *testing.T) {
	_, err := Event("auth", "login")
	if err == nil {
		t.Fatal("expected error for 2 segments")
	}
	var schemaErr *EventSchemaError
	if !errors.As(err, &schemaErr) {
		t.Errorf("expected *EventSchemaError, got %T: %v", err, err)
	}
}

func TestEvent_FiveArgs_Error(t *testing.T) {
	_, err := Event("auth", "login", "user", "detail", "success")
	if err == nil {
		t.Fatal("expected error for 5 segments")
	}
	var schemaErr *EventSchemaError
	if !errors.As(err, &schemaErr) {
		t.Errorf("expected *EventSchemaError, got %T: %v", err, err)
	}
}

func TestEvent_InvalidSegment_StrictMode(t *testing.T) {
	_strictSchema = true
	t.Cleanup(func() { _strictSchema = false })

	_, err := Event("Auth", "login", "success")
	if err == nil {
		t.Fatal("expected error for uppercase segment in strict mode")
	}
	var schemaErr *EventSchemaError
	if !errors.As(err, &schemaErr) {
		t.Errorf("expected *EventSchemaError, got %T: %v", err, err)
	}
	if !strings.Contains(err.Error(), "Auth") {
		t.Errorf("error %q should mention invalid segment %q", err.Error(), "Auth")
	}
}

func TestEvent_InvalidSegment_NonStrictMode(t *testing.T) {
	_strictSchema = false
	evt, err := Event("Auth", "Login", "Success")
	if err != nil {
		t.Fatalf("unexpected error in non-strict mode: %v", err)
	}
	if evt.Event != "Auth.Login.Success" {
		t.Errorf("want %q, got %q", "Auth.Login.Success", evt.Event)
	}
}

func TestEventRecord_Attrs_DAS(t *testing.T) {
	evt := EventRecord{
		Event:  "auth.login.success",
		Domain: "auth",
		Action: "login",
		Status: "success",
	}
	attrs := evt.Attrs()
	if len(attrs) != 3 {
		t.Fatalf("want 3 attrs for DAS, got %d", len(attrs))
	}
	type kv struct{ key, val string }
	want := []kv{
		{"event.domain", "auth"},
		{"event.action", "login"},
		{"event.status", "success"},
	}
	for i, a := range attrs {
		sa, ok := a.(slog.Attr)
		if !ok {
			t.Fatalf("attr[%d] is not slog.Attr: %T", i, a)
		}
		if sa.Key != want[i].key {
			t.Errorf("attr[%d] key: want %q, got %q", i, want[i].key, sa.Key)
		}
		if sa.Value.String() != want[i].val {
			t.Errorf("attr[%d] val: want %q, got %q", i, want[i].val, sa.Value.String())
		}
	}
}

func TestValidateRequiredKeys_AllPresent(t *testing.T) {
	attrs := map[string]any{"domain": "user", "action": "login"}
	err := ValidateRequiredKeys(attrs, []string{"domain", "action"})
	if err != nil {
		t.Errorf("expected no error, got %v", err)
	}
}

func TestValidateRequiredKeys_MissingKey(t *testing.T) {
	attrs := map[string]any{"domain": "user"}
	err := ValidateRequiredKeys(attrs, []string{"domain", "action"})
	if err == nil {
		t.Error("expected error for missing required key 'action'")
	}
}

func TestValidateRequiredKeys_EmptyRequired(t *testing.T) {
	attrs := map[string]any{"domain": "user"}
	err := ValidateRequiredKeys(attrs, nil)
	if err != nil {
		t.Errorf("expected no error with nil required keys, got %v", err)
	}
}

func TestEventRecord_Attrs_DARS(t *testing.T) {
	evt := EventRecord{
		Event:    "payment.subscription.renewal.success",
		Domain:   "payment",
		Action:   "subscription",
		Resource: "renewal",
		Status:   "success",
	}
	attrs := evt.Attrs()
	if len(attrs) != 4 {
		t.Fatalf("want 4 attrs for DARS, got %d", len(attrs))
	}
	found := false
	for _, a := range attrs {
		sa, ok := a.(slog.Attr)
		if !ok {
			t.Fatalf("attr element is not slog.Attr: %T", a)
		}
		if sa.Key == "event.resource" && sa.Value.String() == "renewal" {
			found = true
		}
	}
	if !found {
		t.Error("expected event.resource=renewal in attrs")
	}
}
